### PR TITLE
Added optionsFunction to allow for dynamic options

### PR DESCRIPTION
**Pull request by corbettbarr**
In multi-tenant environments, the optionsFunction will be called during passport.authenticate so options can be created dynamically based on incoming request.

optionsFunction should be provided in constructor options object, and return authorizationURL, tokenURL and clientID. callbackURL and sessionKey must be provided as static values in options object.

### CODE REVIEW

**Comment by corbettbarr** [comment]
No @Bnaya I haven't released it as a package, but feel free to do so if you'd like. This PR is a fairly small set of changes.

**Comment by brodly** [comment]
Whats the status on merging this PR? Would love to use this feature... currently using it on SAML and need similar functionality for Oauth.

File: test/oauth2.test.js [hunk 1]
```diff
@@ -90,7 +90,53 @@ describe('OAuth2Strategy', function() {
         }).to.throw(TypeError, 'OAuth2Strategy requires a authorizationURL option');
       });
     }); // with only a verify callback
-    
+
+    describe('with an optionsFunction and no sessionKey', function() {
+      it('should throw', function() {
+        expect(function() {
+          new OAuth2Strategy({
+            optionsFunction: () => {
```
Comment (brodly) [inline test/oauth2.test.js:98]:
```suggestion
            optionsFunction: function() { 
```

remove es6 syntax for <= node 3

File: test/oauth2.test.js [hunk 2]
```diff
@@ -90,7 +90,53 @@ describe('OAuth2Strategy', function() {
...
+          }, function() {});
+        }).to.throw();
+      })
+    }); // with an optionsFunction and no sessionKey
+
+    describe('with an optionsFunction and sessionKey', function() {
+      it('should not throw', function() {
+        expect(function() {
+          new OAuth2Strategy({
+            optionsFunction: () => {
```
Comment (brodly) [inline test/oauth2.test.js:115]:
```suggestion
            optionsFunction: function() { 
```
remove es6 syntax

File: test/oauth2.test.js [hunk 3]
```diff
@@ -447,7 +493,130 @@ describe('OAuth2Strategy', function() {
         expect(url).to.equal('https://www.example.com/oauth2/authorize?foo=bar&state=foo123&response_type=code&redirect_uri=https%3A%2F%2Fwww.example.net%2Fauth%2Fexample%2Fcallback&client_id=ABC123');
       });
     }); // that redirects to authorization server using authorization endpoint that has query parameters including state with state option
+
+    describe('that uses an optionsFunction that does not return authorizationURL', function() {
+      var strategy = new OAuth2Strategy({
+        optionsFunction: (req, cb) => {
```
Comment (brodly) [inline test/oauth2.test.js:499]:
```suggestion
        optionsFunction: function(req, cb) { 
```

File: test/oauth2.test.js [hunk 4]
```diff
@@ -447,7 +493,130 @@ describe('OAuth2Strategy', function() {
...
+
+      it('should error', function() {
+        expect(err).to.be.an.instanceof(Error);
+        expect(err.message).to.equal('optionsFunction must return authorizationURL');
+      });
+    }); // that uses an optionsFunction that does not return authorizationURL
+
+    describe('that uses an optionsFunction that does not return tokenURL', function() {
+      var strategy = new OAuth2Strategy({
+        optionsFunction: (req, cb) => {
```
Comment (brodly) [inline test/oauth2.test.js:529]:
```suggestion
        optionsFunction: function(req, cb) { 
```

File: test/oauth2.test.js [hunk 5]
```diff
@@ -447,7 +493,130 @@ describe('OAuth2Strategy', function() {
...
+
+      it('should error', function() {
+        expect(err).to.be.an.instanceof(Error);
+        expect(err.message).to.equal('optionsFunction must return tokenURL');
+      });
+    }); // that uses an optionsFunction that does not return tokenURL
+
+    describe('that uses an optionsFunction that does not return clientID', function() {
+      var strategy = new OAuth2Strategy({
+        optionsFunction: (req, cb) => {
```
Comment (brodly) [inline test/oauth2.test.js:559]:
```suggestion
        optionsFunction: function(req, cb) { 
```

File: test/oauth2.test.js [hunk 6]
```diff
@@ -447,7 +493,130 @@ describe('OAuth2Strategy', function() {
...
+
+      it('should error', function() {
+        expect(err).to.be.an.instanceof(Error);
+        expect(err.message).to.equal('optionsFunction must return clientID');
+      });
+    }); // that uses an optionsFunction that does not return clientID
+
+    describe('that uses an optionsFunction', function() {
+      var strategy = new OAuth2Strategy({
+        optionsFunction: (req, cb) => {
```
Comment (brodly) [inline test/oauth2.test.js:589]:
```suggestion
        optionsFunction: function(req, cb) { 
```

**Review by brodly — CHANGES_REQUESTED** [review]
Removed es6 functions as they wont compile for < node 3 builds

**Comment by jaredhanson** [comment]
I'm declining to merge this PR, as this functionality is already achievable at the `passport` level and duplicating `optionsFunction` functionality across protocol-specific strategies (OAuth 2.0, SAML, OIDC, etc) would be redundant.

To achieve this at the passport level, rather than statically `use()`ing strategies ahead of time, create them dynamically during request processing.   For example:

```js
app.get('/login/oauth2', function(req, res, next) {
  var strategy = createStrategy(req.query.provider);
  passport.authenticate(strategy)(req, res, next);
});

function createStrategy(provider) {
  var clientID, clientSecret; // Load these from config store based on provider.

  return new OAuth2Strategy({
      clientID: clientID,
      clientSecret: clientSecret,
      callbackURL: '/oauth2/callback',
    },
    function(accessToken, refreshToken, profile, cb) {
      // ...
    });
}
```

More information can be found on the blog post announcing the feature: [Authenticate using Strategy Instances](https://medium.com/passportjs/authenticate-using-strategy-instances-49e58d96ec8c)

**Pull request closed without merging**